### PR TITLE
docs for pnpm fetch

### PR DESCRIPTION
**Pull request by thynson**

### CODE REVIEW

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+You may have read the [official guide] to writing a Dockerfile for a Node.js
+app, if you didn't read it yet, you may want to read it first.
+
+From that guide, we learn to write an optimized Dockerfile for projects using
+pnpm, which shall looks like
+
+```Dockerfile
+FROM node:14
+
+RUN curl -L https://unpkg.com/@pnpm/self-installer | node
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:17]:
```suggestion
RUN curl -f https://get.pnpm.io/v6.js | node - add --global pnpm
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+# Files required by pnpm install
+COPY .npmrc package.json pnpm-lock.yaml pnpmfile.js ./
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:20]:
```suggestion
COPY .npmrc package.json pnpm-lock.yaml .pnpmfile.cjs ./
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+RUN pnpm install --frozon-lockfile --prod
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:22]:
```suggestion
RUN pnpm install --frozen-lockfile --prod
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+title: Boost docker build with pnpm fetch
+---
+
+`pnpm fetch` is a new command introduced in pnpm v6 to boost docker build.
+
+You may have read the [official guide] to writing a Dockerfile for a Node.js
+app, if you didn't read it yet, you may want to read it first.
+
+From that guide, we learn to write an optimized Dockerfile for projects using
+pnpm, which shall looks like
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:12]:
```suggestion
pnpm, which shall look like
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+RUN pnpm install --frozon-lockfile --prod
+
+# Bundle app source
+COPY . .
+
+EXPOSE 8080
+CMD [ "node", "server.js" ]
+```
+
+As long as there is no chang to `.npmrc`, `package.json`, `pnpm-lock.yaml`,
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:31]:
```suggestion
As long as there is no change to `.npmrc`, `package.json`, `pnpm-lock.yaml`,
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+`pnpmfile.js`, docker build cache is still valid up to the layer of
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:32]:
```suggestion
`.pnpmfile.cjs`, docker build cache is still valid up to the layer of
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+contain the version number, scripts, and arbitrary configuration for any other
+tool.
+
+It's also hard to maintain a Dockerfile that build a monorepo project, it may
+look like
+
+```Dockerfile
+FROM node:14
+
+RUN curl -L https://unpkg.com/@pnpm/self-installer | node
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:47]:
```suggestion
RUN curl -f https://get.pnpm.io/v6.js | node - add --global pnpm
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+# Files required by pnpm install
+COPY .npmrc package.json pnpm-lock.yaml pnpmfile.js ./
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:50]:
```suggestion
COPY .npmrc package.json pnpm-lock.yaml .pnpmfile.cjs ./
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+# for each sub-package, we have to add one extra step to copy its manifest
+# to right place, as docker have no way to filter out only package.json with
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:53]:
```suggestion
# to the right place, as docker have no way to filter out only package.json with
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+# single instruction
+COPY packages/foo/backage.json packages/foo/
+COPY packages/bar/backage.json packages/bar/
+
+RUN pnpm install --frozon-lockfile --prod
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:58]:
```suggestion
RUN pnpm install --frozen-lockfile --prod
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+RUN curl -L https://unpkg.com/@pnpm/self-installer | node
+
+# pnpm fetch does require only lockfile
+COPY pnpm-lock.yaml ./
+
+RUN pnpm fetch --prod
+
+
+ADD . ./
+RUN pnpm recursive install --offline --prod
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:85]:
```suggestion
RUN pnpm install -r --offline --prod
```

but actually, in a workspace `pnpm install` is always `pnpm install -r`, so if it is in a workspace, just run `pnpm install`.

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+As you can see, the Dockerfile has to be updated when you add or remove
+sub-packages.
+
+`pnpm fetch` will solve the above problem perfectly by providing the ability
+to fetch package to virtual store with information only from a lockfile.
+
+```Dockerfile
+FROM node:14
+
+RUN curl -L https://unpkg.com/@pnpm/self-installer | node
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:76]:
```suggestion
RUN curl -f https://get.pnpm.io/v6.js | node - add --global pnpm
```

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+EXPOSE 8080
+CMD [ "node", "server.js" ]
+```
+
+It works for both simple project and monorepo project, `--offline` enforces
+pnpm not to communicate with package registry as all needed packages shall be
+already presented in the virtual store.
+
+As long as the lockfile is not changed, the build cache is valid up to the
+layer `RUN pnpm recursive install --offline --prod`, which will save you much
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:99]:
```suggestion
layer `RUN pnpm install -r --offline --prod`, which will save you much
```

File: docs/cli/fetch.md [hunk 14]
```diff
@@ -0,0 +1,22 @@
+---
+id: fetch
+title: pnpm fetch
+---
+
+Fetch packages from a lockfile into virtual store, package manifest is ignored.
+
+WARNING! This is an experimental command. Breaking changes may be introduced in non-major versions of the CLI.
```
Comment (zkochan) [inline docs/cli/fetch.md:8]:
```suggestion
:::warning

This is an experimental command. Breaking changes may be introduced in non-major versions of the CLI.

:::
```

**Review by zkochan — CHANGES_REQUESTED** [review]
I am not sure there is need in two pages, maybe just move everything to the commands doc page

**Comment by thynson** [comment]
Maybe add a link in the fetch command doc to the advanced topic? 
I made it two page because I saw that all command line doc pages are relative simple.

**Comment by zkochan** [comment]
They are simple only because I was lazy to write more docs. This command has a very specific usage scenario, so better to describe it on the command description page.

File: docs/boost-docker-build-with-pnpm-fetch.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+From that guide, we learn to write an optimized Dockerfile for projects using
+pnpm, which shall look like
+
+```Dockerfile
+FROM node:14
+
+RUN curl -f https://get.pnpm.io/v6.js | node - add --global pnpm
+
+# Files required by pnpm install
+COPY .npmrc package.json pnpm-lock.yaml pnpmfile.js ./
```
Comment (zkochan) [inline docs/boost-docker-build-with-pnpm-fetch.md:20]: pnpmfile.js is renamed in pnpm v6

**Review by zkochan — APPROVED** [review]

**Comment by thynson** [comment]
I have to go to sleep now 😴 , and I'll have a trip next few days, so it may be delayed if any additional work is required. 
Thanks for getting this feature merged.

**Comment by zkochan** [comment]
Good night and have a good trip!
Thanks for your contributions!

**Pull request merged**